tcms/issuetracker/azure_boards.py

```diff
@@ -0,0 +1,153 @@
+# -*- coding: utf-8 -*-
+from requests.auth import HTTPBasicAuth
+import requests
+
+from tcms.core.contrib.linkreference.models import LinkReference
+from tcms.issuetracker.base import IssueTrackerType, IntegrationThread
+
+
+class AzureAPI():
+    """
+        Azure Boards API interaction class.
+
+    """
+    def __init__(self, base_url=None, password=None):
+        self.api_version = "?api-version=6.0"
+        self.headers = {"Accept": 'application/json-patch+json',
+                        'Content-type': 'application/json-patch+json'}
+        self.auth = HTTPBasicAuth('apikey', password)
+        self.base_url = base_url + "/_apis/"
+
+    def get_issue(self, issue_id):
+        url = "{0}{1}{2}{3}".format(self.base_url, "wit/workitems/",
+                                    str(issue_id), self.api_version)
+        return self._get(url, self.headers, self.auth)
+
+    def create_issue(self, body):
+        url = "{0}{1}{2}".format(self.base_url, "wit/workitems/$Issue", self.api_version)
+        return self._post(url, self.headers, self.auth, body)
+
+    def update_issue(self, issue_id, body):
+        url = "{0}{1}{2}{3}".format(self.base_url, "wit/workitems/",
+                                    str(issue_id), self.api_version)
+        return self._patch(url, self.headers, self.auth, body)
+
+    def add_comment(self, issue_id, body):
+        headers = {"Content-type": "application/json"}
+        url = "{0}{1}{2}{3}{4}{5}".format(self.base_url, "wit/workitems/",
+                                          str(issue_id), "/comments", self.api_version,
+                                          "-preview.3")
+        return self._post(url, headers, self.auth, body)
+
+    def get_comments(self, issue_id):
+        headers = {"Content-type": "application/json"}
+        url = "{0}{1}{2}{3}{4}{5}".format(self.base_url, "wit/workitems/",
+                                          str(issue_id), "/comments", self.api_version,
+                                          "-preview.3")
+        return self._get(url, headers, self.auth)
+
+    @staticmethod
+    def _get(url, headers, auth):
+        response = requests.get(url, headers=headers, auth=auth)
+        return response.json()
+
+    @staticmethod
+    def _post(url, headers, auth, body):
+        response = requests.post(url, headers=headers, auth=auth, json=body)
+        return response.json()
+
+    @staticmethod
+    def _patch(url, headers, auth, body):
+        response = requests.patch(url, headers=headers, auth=auth, json=body)
+        return response.json()
+
+
+class AzureThread(IntegrationThread):
+    """
+        Execute AzureBoards code in a thread!
+
+        Executed from the IssueTracker interface methods.
+    """
+    def post_comment(self):
+        # NOTE: Posting comment is in preview state in API v6.0.
+        # Throws internal error.
+        comment_body = {
+                "text": self.text().replace("\n", "<br/>")
```

Comment:
Is HTML allowed inside Azure Boards comments ? What happens to the newlines if you don't remove them?

Reply:
Both MD and HTML are supported and if line endings are not replaced with "<br/>" then lines get concatenated.

See section "Add guidance in a rich-text field";
https://docs.microsoft.com/en-us/azure/devops/boards/work-items/work-item-template-examples?view=azure-devops

Reply:
Couple of notes:
- this should end up as a comment in the source code + the URL reference. 
- the docs have a permalink to the section you are refering to
- I see in the docs that HTML is supported but they don't say anything about newlines

But more importantly on our side everything is markdown so I'm wondering how that would be rendered on the Azure side if we post everything as-is! If Azure really uses HTML internally we could probably do a better job at converting to HTML before posting to Azure.

Reply:
So tried out at https://dev.azure.com/kiwitcms/boards-integration/_workitems/edit/6/ and it looks like Markdown isn't supported but HTML is. Also any of the other issues coming from tests doesn't look nice too:
https://dev.azure.com/kiwitcms/boards-integration/_workitems/edit/4/

Instead of replacing newlines go directly for the `markdown2html()` function which we've got.

Reply:
markdown2html working nicely. Compare issue#4 from the previous comment to https://dev.azure.com/kiwitcms/boards-integration/_workitems/edit/13/